src/Tooltip/Tooltip.js

```diff
@@ -0,0 +1,141 @@
+import classnames from 'classnames';
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+const METHODS_TO_BIND = [
+  'getIdealPosition',
+  'handleMouseEnter',
+  'handleMouseLeave'
+];
+
+class Tooltip extends React.Component {
+  constructor() {
+    super();
+
+    this.state = {isOpen: false};
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  getIdealPosition() {
+    // Get the anchor and position from state if possible. If not, get it from
+    // the props.
+    let anchor = this.state.anchor || this.props.anchor;
+    let position = this.state.position || this.props.position;
+
+    if (this.refs.tooltipContent) {
+      let viewportHeight = global.window.innerHeight;
+      let viewportWidth = global.window.innerWidth;
+      let tooltipContent = ReactDOM.findDOMNode(this.refs.tooltipContent);
+      let tooltipPosition = tooltipContent.getBoundingClientRect();
+
+      // Change the position if the tooltip will be rendered off the screen.
+      if (position === 'right' && tooltipPosition.right > viewportWidth) {
+        position = 'left';
+      } else if (position === 'left' && tooltipPosition.left < 0) {
+        position = 'right';
+      } else if (position === 'top' && tooltipPosition.top < 0) {
+        position = 'bottom';
+      } else if (position === 'bottom' &&
+        tooltipPosition.bottom > viewportHeight) {
+        position = 'top';
+      }
+
+      // Change the anchor if the tooltip will be rendered off the screen.
+      if (position === 'right' || position === 'left') {
+        if (tooltipPosition.top < 0) {
+          anchor = 'start';
+        } else if (tooltipPosition.bottom > viewportHeight) {
+          anchor = 'end';
+        }
+      } else if (position === 'top' || position === 'bottom') {
+        if (tooltipPosition.right > viewportWidth) {
+          anchor = 'end';
+        } else if (tooltipPosition.left < 0) {
+          anchor = 'start';
+        }
+      }
+    }
+
+    return {anchor, position};
+  }
+
+  handleMouseEnter() {
+    let {anchor, position} = this.getIdealPosition();
+    this.setState({anchor, isOpen: true, position});
+  }
+
+  handleMouseLeave() {
+    this.setState({isOpen: false});
+  }
+
+  render() {
+    let {props, state} = this;
+
+    // Get the anchor and position from state if possible. If not, get it from
+    // the props.
+    let anchor = state.anchor || props.anchor;
```

Comment:
We can't use any function in `#render` that calls `#findDOMNode`, see the error from React:
![](http://cl.ly/3X2b3y3S000L/Screen%20Shot%202016-02-03%20at%2011.55.23%20AM.png)